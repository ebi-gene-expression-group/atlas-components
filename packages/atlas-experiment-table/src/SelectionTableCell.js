import React from 'react'
import PropTypes from 'prop-types'

import { Table, Checkbox } from 'evergreen-ui'

const SelectionTableCell = ({ dataRow, dataKey, selectedRowIds, onChange, width }) =>
  <Table.Cell justifyContent={`center`} flexGrow={width}>
    <Checkbox
      checked={selectedRowIds.includes(dataRow[dataKey])}
      onChange={() => onChange(dataRow[dataKey])}/>
  </Table.Cell>

SelectionTableCell.propTypes = {
  dataRow: PropTypes.object.isRequired,
  dataKey: PropTypes.string.isRequired,
  selectedRowIds: PropTypes.arrayOf(PropTypes.string).isRequired,
  onChange: PropTypes.func,
  width: PropTypes.number
}

SelectionTableCell.defaultProps = {
  width: 1
}

export default SelectionTableCell
