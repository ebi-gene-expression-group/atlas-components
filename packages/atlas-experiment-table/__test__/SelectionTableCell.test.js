import React from 'react'
import { shallow, mount } from 'enzyme'

import { Table, Checkbox } from 'evergreen-ui'

import SelectionTableCell from '../src/SelectionTableCell'

import randomString from 'random-string'

describe(`SelectionTableCell`, () => {
  const props = {
    dataRow: {
      string: randomString()
    },
    dataKey: `string`,
    selectedRowIds: [],
    selectOnChange: jest.fn(),
  }

  test(`renders a table cell with a checkbox`, () => {
    const wrapper = shallow(<SelectionTableCell {...props}/>)

    expect(wrapper.find(Table.Cell).find(Checkbox)).toHaveLength(1)
  })

  test(`displays an unchecked checkbox if the current row is not in the array of selected rows`, () => {
    const wrapper = shallow(<SelectionTableCell {...props} selectedRowIds={[randomString()]}/>)

    expect(wrapper.find(Checkbox)).toHaveProp({checked: false})
  })

  test(`displays a checked checkbox if the current row is in the array of selected rows`, () => {
    const wrapper = shallow(<SelectionTableCell {...props} selectedRowIds={[props.dataRow[props.dataKey]]}/>)

    expect(wrapper.find(Checkbox)).toHaveProp({checked: true})
  })

  test(`calls selectOnChange with the keyed value from dataRow as argument when the checkbox is (de)selected`, () => {
    const wrapper = shallow(<SelectionTableCell {...props}/>)
    wrapper.find(Checkbox).simulate(`change`)

    expect(props.selectOnChange).toHaveBeenCalled()
    expect(props.selectOnChange.mock.calls).toContainEqual([props.dataRow[props.dataKey]])
  })

  test(`matches snapshot (unchecked)`, () => {
    const props = {
      dataRow: {
        string: `foobar`
      },
      dataKey: `string`,
      selectedRowIds: [],
      selectOnChange: jest.fn(),
    }
    const wrapper = mount(<SelectionTableCell {...props}/>)
    expect(wrapper).toMatchSnapshot()
  })

  test(`matches snapshot (checked)`, () => {
    const props = {
      dataRow: {
        string: `foobar`
      },
      dataKey: `string`,
      selectedRowIds: [`foobar`],
      selectOnChange: jest.fn(),
    }
    const wrapper = mount(<SelectionTableCell {...props}/>)
    expect(wrapper).toMatchSnapshot()
  })
})
