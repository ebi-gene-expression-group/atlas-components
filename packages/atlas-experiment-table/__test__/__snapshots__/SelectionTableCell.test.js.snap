// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectionTableCell matches snapshot (checked) 1`] = `
<SelectionTableCell
  dataKey="string"
  dataRow={
    Object {
      "string": "foobar",
    }
  }
  selectOnChange={[MockFunction]}
  selectedRowIds={
    Array [
      "foobar",
    ]
  }
  width={1}
>
  <ForwardRef(TableCell)
    flexGrow={1}
    justifyContent="center"
  >
    <ForwardRef(Pane)
      alignItems="center"
      boxSizing="border-box"
      className="css-saeo60"
      display="flex"
      flex={1}
      flexGrow={1}
      flexShrink={0}
      justifyContent="center"
      onKeyDown={[Function]}
      overflow="hidden"
      paddingX={12}
    >
      <Box
        alignItems="center"
        boxSizing="border-box"
        className="css-saeo60 css-nil"
        display="flex"
        flex={1}
        flexGrow={1}
        flexShrink={0}
        is="div"
        justifyContent="center"
        onKeyDown={[Function]}
        overflow="hidden"
        paddingX={12}
      >
        <div
          className="css-saeo60 css-nil ub-pl_12px ub-pr_12px ub-box-szg_border-box ub-flx_1 ub-dspl_flex ub-algn-itms_center ub-flx-srnk_0 ub-ovflw-x_hidden ub-ovflw-y_hidden ub-just-cnt_center ub-flx-grow_1"
          onKeyDown={[Function]}
        >
          <ForwardRef(Checkbox)
            checked={true}
            onChange={[Function]}
          >
            <Box
              boxSizing="border-box"
              cursor="pointer"
              display="flex"
              is="label"
              marginY={16}
              position="relative"
            >
              <label
                className="ub-crsr_pointer ub-pst_relative ub-dspl_flex ub-mb_16px ub-mt_16px ub-box-szg_border-box"
              >
                <Box
                  boxSizing="border-box"
                  checked={true}
                  className="css-1bkiqpw"
                  is="input"
                  onChange={[Function]}
                  type="checkbox"
                >
                  <input
                    checked={true}
                    className="css-1bkiqpw ub-box-szg_border-box"
                    onChange={[Function]}
                    type="checkbox"
                  />
                </Box>
                <Box
                  alignItems="center"
                  borderRadius={3}
                  boxSizing="border-box"
                  display="flex"
                  flex="none"
                  height={16}
                  is="div"
                  justifyContent="center"
                  width={16}
                >
                  <div
                    className="ub-box-szg_border-box ub-bblr_3px ub-bbrr_3px ub-btlr_3px ub-btrr_3px ub-dspl_flex ub-flx_none ub-algn-itms_center ub-just-cnt_center ub-w_16px ub-h_16px"
                  >
                    <CheckIcon>
                      <svg
                        height={7}
                        viewBox="0 0 10 7"
                        width={10}
                      >
                        <path
                          d="M4 4.586L1.707 2.293A1 1 0 1 0 .293 3.707l3 3a.997.997 0 0 0 1.414 0l5-5A1 1 0 1 0 8.293.293L4 4.586z"
                          fill="currentColor"
                          fillRule="evenodd"
                        />
                      </svg>
                    </CheckIcon>
                  </div>
                </Box>
              </label>
            </Box>
          </ForwardRef(Checkbox)>
        </div>
      </Box>
    </ForwardRef(Pane)>
  </ForwardRef(TableCell)>
</SelectionTableCell>
`;

exports[`SelectionTableCell matches snapshot (unchecked) 1`] = `
<SelectionTableCell
  dataKey="string"
  dataRow={
    Object {
      "string": "foobar",
    }
  }
  selectOnChange={[MockFunction]}
  selectedRowIds={Array []}
  width={1}
>
  <ForwardRef(TableCell)
    flexGrow={1}
    justifyContent="center"
  >
    <ForwardRef(Pane)
      alignItems="center"
      boxSizing="border-box"
      className="css-saeo60"
      display="flex"
      flex={1}
      flexGrow={1}
      flexShrink={0}
      justifyContent="center"
      onKeyDown={[Function]}
      overflow="hidden"
      paddingX={12}
    >
      <Box
        alignItems="center"
        boxSizing="border-box"
        className="css-saeo60 css-nil"
        display="flex"
        flex={1}
        flexGrow={1}
        flexShrink={0}
        is="div"
        justifyContent="center"
        onKeyDown={[Function]}
        overflow="hidden"
        paddingX={12}
      >
        <div
          className="css-saeo60 css-nil ub-pl_12px ub-pr_12px ub-box-szg_border-box ub-flx_1 ub-dspl_flex ub-algn-itms_center ub-flx-srnk_0 ub-ovflw-x_hidden ub-ovflw-y_hidden ub-just-cnt_center ub-flx-grow_1"
          onKeyDown={[Function]}
        >
          <ForwardRef(Checkbox)
            checked={false}
            onChange={[Function]}
          >
            <Box
              boxSizing="border-box"
              cursor="pointer"
              display="flex"
              is="label"
              marginY={16}
              position="relative"
            >
              <label
                className="ub-crsr_pointer ub-pst_relative ub-dspl_flex ub-mb_16px ub-mt_16px ub-box-szg_border-box"
              >
                <Box
                  boxSizing="border-box"
                  checked={false}
                  className="css-1bkiqpw"
                  is="input"
                  onChange={[Function]}
                  type="checkbox"
                >
                  <input
                    checked={false}
                    className="css-1bkiqpw ub-box-szg_border-box"
                    onChange={[Function]}
                    type="checkbox"
                  />
                </Box>
                <Box
                  alignItems="center"
                  borderRadius={3}
                  boxSizing="border-box"
                  display="flex"
                  flex="none"
                  height={16}
                  is="div"
                  justifyContent="center"
                  width={16}
                >
                  <div
                    className="ub-box-szg_border-box ub-bblr_3px ub-bbrr_3px ub-btlr_3px ub-btrr_3px ub-dspl_flex ub-flx_none ub-algn-itms_center ub-just-cnt_center ub-w_16px ub-h_16px"
                  >
                    <CheckIcon>
                      <svg
                        height={7}
                        viewBox="0 0 10 7"
                        width={10}
                      >
                        <path
                          d="M4 4.586L1.707 2.293A1 1 0 1 0 .293 3.707l3 3a.997.997 0 0 0 1.414 0l5-5A1 1 0 1 0 8.293.293L4 4.586z"
                          fill="currentColor"
                          fillRule="evenodd"
                        />
                      </svg>
                    </CheckIcon>
                  </div>
                </Box>
              </label>
            </Box>
          </ForwardRef(Checkbox)>
        </div>
      </Box>
    </ForwardRef(Pane)>
  </ForwardRef(TableCell)>
</SelectionTableCell>
`;
